src/main/java/org/hawkular/client/android/activity/AuthorizationActivity.java

```diff
@@ -0,0 +1,210 @@
+            hostEdit.append(BackendEndpoints.Community.HOST);
+            portEdit.append(BackendEndpoints.Community.PORT);
+        }
+    }
+
+    @OnClick(R.id.button_authorize)
+    public void setUpAuthorization() {
+        if (getHost().isEmpty()) {
+            showError(hostEdit, R.string.error_empty);
+            return;
+        }
+
+        if (getPort().isEmpty()) {
+            showError(portEdit, R.string.error_empty);
+            return;
+        }
+
+        setUpBackendAuthorization();
+    }
+
+    private String getHost() {
+        return hostEdit.getText().toString().trim();
+    }
+
+    private String getPort() {
+        return portEdit.getText().toString().trim();
+    }
+
+    private void showError(EditText errorEdit, @StringRes int errorMessage) {
+        errorEdit.setError(getString(errorMessage));
+    }
+
+    private void setUpBackendAuthorization() {
+        try {
+            BackendClient.getInstance().setUpBackend(getHost(), getPort());
+
+            BackendClient.getInstance().authorize(this, this);
+        } catch (RuntimeException e) {
+            Timber.d(e, "Authorization failed.");
+
+            showError(R.string.error_authorization_host_port);
+        }
+    }
+
+    private void showError(@StringRes int errorMessage) {
+        Snackbar.make(findViewById(android.R.id.content), errorMessage, Snackbar.LENGTH_LONG).show();
+    }
+
+    @Override
+    public void onSuccess(String authorization) {
+        setUpTenant();
+    }
+
+    @Override
+    public void onFailure(Exception e) {
+        Timber.d(e, "Authorization failed.");
+
+        showError(R.string.error_general);
+    }
+
+    private void setUpTenant() {
+        BackendClient.getInstance().getTenants(this, new TenantsCallback());
+    }
+
+    private void setUpEnvironment(Tenant tenant) {
+        BackendClient.getInstance().getEnvironments(tenant, this, new EnvironmentsCallback(tenant));
+    }
+
+    private void succeed(Tenant tenant, Environment environment) {
+        saveBackendPreferences(tenant, environment);
+
+        setResult(Activity.RESULT_OK);
+        finish();
+    }
+
+    private void saveBackendPreferences(Tenant tenant, Environment environment) {
+        Preferences.ofBackend(this).host().set(getHost());
+        Preferences.ofBackend(this).port().set(getPort());
+        Preferences.ofBackend(this).tenant().set(tenant.getId());
+        Preferences.ofBackend(this).environment().set(environment.getId());
+    }
+
+    private static final class TenantsCallback extends AbstractActivityCallback<List<Tenant>> {
+        @Override
+        public void onSuccess(List<Tenant> tenants) {
+            if (tenants.isEmpty()) {
+                Timber.d("Tenants list is empty, this should not happen.");
+                return;
+            }
+
+            // This is a potentially dangerous action.
+            // AeroGear does not support single item fetching.
+            Tenant tenant = tenants.get(0);
```

Comment:
Probably not. I don’t have the ID at this point, so the only thing available is `Callback<List<T>>`, even if the response is `{"key": "value"}` instead of `[{"key": "value}]`.